### PR TITLE
Profile batch transfer and gradient clipping hooks

### LINKED ISSUE #14028: Some `LightningModule`/`LightningDataModule` hooks are not profiled
## 🐛 Bug

These hooks are not profiled, which I think should be profiled:
```console
'on_before_batch_transfer'
'transfer_batch_to_device'
'on_after_batch_transfer'
'configure_gradient_clipping'
```

### Environment

- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): LightningModule
- PyTorch Lightning Version (e.g., 1.5.0): master
- Lightning App Version (e.g., 0.5.2):
- PyTorch Version (e.g., 1.10):
- Python version (e.g., 3.9):
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Running environment of LightningApp (e.g. local, cloud):
- Any other relevant information:

### Additional context

cc @carmocca @kaushikb11 @ninginthecloud @rohitgr7 @nbcsm @guotuofeng

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #14028

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @kaushikb11 @ninginthecloud @rohitgr7 @nbcsm @guotuofeng @awaelchli @borda @akihironitta

### CODE REVIEW

**Comment by rohitgr7** [comment]
any suggestions for the test here?

I am thinking of extending these with profilers since they cover all the cases with hooks.
https://github.com/Lightning-AI/lightning/blob/master/tests/tests_pytorch/models/test_hooks.py

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -291,16 +290,24 @@ def _apply_batch_transfer_handler(
...
+                    else self._trainer._call_lightning_datamodule_hook
+                )
+                return trainer_method(hook_name, **kwargs)
+            else:
+                hook = getattr(self, hook_name)
+                return hook(**kwargs)
+
+        batch = call_hook("on_before_batch_transfer", batch=batch, dataloader_idx=dataloader_idx)
+        batch = call_hook("transfer_batch_to_device", batch=batch, device=device, dataloader_idx=dataloader_idx)
+        batch = call_hook("on_after_batch_transfer", batch=batch, dataloader_idx=dataloader_idx)
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:310]: I generally prefer to pass positional arguments positionally as it was done before. This will also avoid all those test updates.

File: src/pytorch_lightning/plugins/precision/precision_plugin.py [hunk 2]
```diff
@@ -182,9 +182,11 @@ def _clip_gradients(
...
             # the configuration validator disallows clipping on manual
             return
-        model.configure_gradient_clipping(
-            optimizer,
-            optimizer_idx,
+
+        model.trainer._call_lightning_module_hook(
+            "configure_gradient_clipping",
+            optimizer=optimizer,
+            optimizer_idx=optimizer_idx,
```
Comment (carmocca) [inline src/pytorch_lightning/plugins/precision/precision_plugin.py:189]: Same comment about positional arguments

File: src/pytorch_lightning/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -545,27 +545,27 @@ class _DataHookSelector:
         default=("on_before_batch_transfer", "transfer_batch_to_device", "on_after_batch_transfer")
     )
 
-    def get_hook(self, hook_name: str) -> Callable:
+    def get_instance(self, hook_name: str) -> Callable:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:548]: @awaelchli might've chosen hook for a specific reason

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:548]: I added this.

File: src/pytorch_lightning/core/module.py (hunk 1 above)
Comment (rohitgr7) [reply src/pytorch_lightning/core/module.py:310]: I think this is less error prone in case some updates are made to the API.
anyways, it doesn't affect the end user

File: src/pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:548]: Nonetheless, the dataclass loses it's meaning. 
Now, the responsibility of mapping the callable is no longer here. You need to update the design and description.

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:548]: The return type needs to be  updated too.

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py [hunk 4]
```diff
@@ -187,11 +187,6 @@ def __init__(self, not_supported):
             {
                 "log",
                 "log_dict",
-                # the following are problematic as they do have `self._current_fx_name` defined some times but
-                # not others depending on where they were called. So we cannot reliably `self.log` in them
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/logging_/test_logger_connector.py:191]: How has this been addressed? Or was it long outdated?

Comment (rohitgr7) [reply tests/tests_pytorch/trainer/logging_/test_logger_connector.py:191]: the batch transfer hooks are not called using `trainer._call_lightningmodule_hook()`, so they will have current_fx_name and no logging is allowed in these hooks as per `logger_connector/fx_validator`.

File: src/pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: updated!

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py (hunk 4 above)
Comment (awaelchli) [reply tests/tests_pytorch/trainer/logging_/test_logger_connector.py:191]: Okay, I wanted to make sure because the comment must have had a meaning at some point :)

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Added the missing chlog entry

**Comment by carmocca** [comment]
Why did we include this in the bug-fix release? Technically this just added a feature (profiling)

**Comment by awaelchli** [comment]
I included it because it was labelled as 1.7.x and also the linked issue had a bug label. This was evidence enough for me that it is correct and intended. If I'm not sure (for example, when linked issue is not also labelled accordingly), I usually ask (https://github.com/Lightning-AI/lightning/pull/9938#issuecomment-1215369656)

**Comment by rohitgr7** [comment]
It was totally unintentional. The hook was missing from the logging `FxValidator` that enabled logging in this hook. The real intention was just to ensure these hooks are profiled, so I added it as a bug, but logging is part of the `trainer._call_lightningmodule_hook`.
